Cogburn/wip module

coreyogburn opened the pull request.

mc-wright committed:
Initial UI prototyping for a Chat page

coreyogburn committed:
New Module Basics

mc-wright committed:
initial api integration

coreyogburn committed:
Barebones Chat

Passthrough to the `/chat` endpoint to the Gateway.

mc-wright committed:
fix

coreyogburn committed:
Barebones Balance

mc-wright committed:
fix

coreyogburn committed:
Configure ResourceManager Properly

mc-wright committed:
working bare bones ai chat interface

mc-wright committed:
added message history support to ai requests

coreyogburn committed:
Improved Chat

ChatWithMessages is the new Chat.

Updated tests to use a generated mock and the assertion library.

Added streaming support to the chat endpoint for any requests with `Accept: text/event-stream` header. If for whatever reason the incoming request doesn't support streaming, the request will fall back to returning all data synchronously.

Regenerated all mocks.

mc-wright committed:
stays on current chat when switching between pages

mc-wright committed:
credits count turns red at 50000 left

mc-wright committed:
fix

coreyogburn committed:
Streaming Chat

We now support streaming! The backend can send up the right headers and content values to have the response streamed back to the user. This requires modifying nginx.conf to work or else nginx will buffer between the backend and the frontend.

The frontend has been refactored to use fetch for the chat request because axios does not seem to support streaming. Processes only `data:` messages and account for messages split across reads. Note: When you await fetch, the promise is fulfilled when the browser gets the headers back.

mc-wright committed:
updated blanceresponse struct

coreyogburn committed:
Tool Usage + First Tool Backend Support

Laid out the foundation for a self documenting FunctionLibrary.

Implemented the first tool: query_events. Borrowed a lot of edge case logic from Mike's prototype.

New endpoint allows the UI to request a tool be executed and added to the chat.

Regenerated mocks.

Added some tests.

TODO: more work around attaching chats to sessions needs to be done.

mc-wright committed:
added initial prototype for investigate alert feature

mc-wright committed:
fixed issue with json chunking

coreyogburn committed:
Tool Execution

Changed our approach to tools. Instead of adding descriptions of tools to the prompt, we're using the API provided mechanisms.

Updated options for Chat so tool use can be automatic or manual in soc. ChatStream will be used for direct user interaction and should always allow the UI to determine if a tool gets executed or not.

Cleaned up some of the documentation contents for our only tool.

coreyogburn committed:
CreditBalance Field Renaming

mc-wright committed:
ui updates for chat screen

coreyogburn committed:
Incremental Progressive for Automatic Tool Usage

Better ToolResult support. When doing a non-streaming chat, tool execution should happen automatically, the results sent to the AI, and the final result sent back.

mc-wright committed:
preliminary frontend tool integration into chat window

coreyogburn committed:
Fix Query

coreyogburn committed:
Include Chat History in Tool Use

PostTool now accepts a body with the full chat history.

coreyogburn committed:
"No Results" is no longer an error

mc-wright committed:
added streaming support for tool use responses

coreyogburn committed:
Saving Chat History

This saves incoming chat requests into elastic.

coreyogburn committed:
Reenable Streaming Support

Reverted a temporary change that shouldn't have been committed.

mc-wright committed:
fixed some visual inconsistencies in the chat screen

coreyogburn committed:
Persist Message History

Save messages sent between the user and the AI. While streaming, we retain the data for later processing. Once the message is fully sent, we parse it and save it.

All messages sent and received in chat and tool usage should now be saved. More work needs to be done in conjunction with the UI to ensure they're added to the proper sessions.

mc-wright committed:
added support for tool use chaining in the chat screen

mc-wright committed:
troubleshooting content block

coreyogburn committed:
Content Oddities

When responding with the tool result, the AI API only respects `content: "string"` despite requesting a ContentBlock.

mc-wright committed:
put chat sessionID in route path

coreyogburn committed:
Writing Chat History

We need to use a different structure for storing a message than we do for transferring one. Models have been updated to account for this.

Parsing and saving a chat message after the client has disconnected needed refactoring. The context was being canceled by the disconnected client so the save wouldn't go through.

Improved validation.

coreyogburn committed:
Get Chat History

Requesting history will give you back the first message of every chat on record. Every message will have a sessionId, a timestamp, and some user content we can use to title the conversation.

mc-wright committed:
progress on AI investigations feature

coreyogburn committed:
GetSessionHistory Endpoint

mc-wright committed:
Integrating elasticsearch chat history into frontend

mc-wright committed:
raw tool result dropdown for tool use

coreyogburn committed:
Tags and DeletedAt on Model, Delete Endpoint

Elasticsearch storage models have been updated to allow tagging, deleting, and the storage of tool usage parameters.

A new endpoint has been added that allows for the deleting of chat sessions. Because sessions are ethereal, special settings are specified on the first message of the session. Calls for previous conversations will not return deleted sessions.

coreyogburn committed:
Fix Parsing of Input on Tool Usage Requests

Input comes is as an object, not a string. So far it's always been an empty object and the updates are always partial json so drop the object and assume an empty string.

coreyogburn committed:
Added Test, Updated Mocks, Wait_For Refresh

coreyogburn committed:
Fix GetPreviousConversations Returning Second Message

The query in GetPreviousConversations was supposed to return "the first message of [sessions that hadn't been deleted]" but was returning "[the first message that hadn't been deleted] from every session". This new query returns what we want.

Reverted the refresh query param in indexDocument.

mc-wright committed:
chat session delete feature and fixed sessionID generation timing

mc-wright committed:
jest testing for chat page

coreyogburn committed:
ES-Backed Message History

Refactored UI to stop building message history and sending it up with chat and tool requests.

Updated models to match what the AI Gateway and AI API expect. Removed a lot of unnecessary complexity.

Updated handlers to retrieve history by sessionId rather than expecting it to be passing in by the request. Updated tests.

Updated AssistantManager interface for non-streaming chat to return all messages that get created (i.e. automatic tool use and tool result messages).

Solved ToolResult. Ultimately we're sending it back to the AI as a user message. This appears to be the only way to get the API to not error and response to a tool result. These messages are now tagged with `tool_result` so the UI can easily identify them.

coreyogburn committed:
Looping ToolRequests

If a non-streaming chat requests tools, check the responses to see if they also request tools. Loop until the final response does not request tools. For now that could be infinite. Killing soc would end this loop.

mc-wright committed:
integrated tool use tags into frontend

mc-wright committed:
tool use completion timestamp support

mc-wright committed:
fix

mc-wright committed:
investigate alerts feature first steps

mc-wright committed:
context length tracking

mc-wright committed:
tool use rejection tracking

coreyogburn committed:
get_playbook_questions tool, more reliable tool results

get_playbook_questions mostly implemented. It fetches the questions and simplifies them down so the investigation can continue. I'll be working to expand this to include the answers to the questions but checking in now so the frontend has something to work with.

The assistanthandler now works harder to ensure that a success or error message is always added to the session.

mc-wright committed:
tool use error handling and display, investigation tooltip update

coreyogburn committed:
get_playbook_questions now serves results

Improved tool so it fetches the answers to the questions.

Updated Investigation prompt to adapt to tool changes.

Added and updated tests.

Eventually, the current back-and-forth playbook conversion (send the playbook to the UI, UI substitutes variables, send the questions to the backend, backend converts and runs queries) that takes place in Alerts will be refactored to use this new backend-only code. But I'm going to leave them separate for a bit as we gain confidence in this new implementation.

mc-wright committed:
increase max context threshold option toggle

mc-wright committed:
context increase option wording update

coreyogburn committed:
Add tests

mc-wright committed:
i18n update for investigated toggle

mc-wright committed:
fix i18n

coreyogburn committed:
License, Permissions, Doc Strings, Fixes

Added OAI feature to the license manager. Updated tests.

Added permissions for new assistant functionality. *_authored grants access to self authored data while *_all will be used by managers to manage data authored by other people. Updated whitespace to keep the columns even.

Permission checks were added to the handlers as well as swagger doc strings. Tests updated for permission checks.

Fixed a bug caught by cypress when creating a case from an event. Certain frontend-only fields were making it to the backend resulting in a 500. hunt.js:770 contains the only functional change made to the file.

coreyogburn committed:
Pass through new Client params

mc-wright committed:
icon changes, refresh bug fix, added much more comprehensive chat page testing

mc-wright committed:
renamed chat pages to assistant

coreyogburn committed:
Api Key from Pro Key, Client Params

No need for an ApiKey in the config, we're building it off of the Pro Key's signature now.

New assistant client parameter.

coreyogburn committed:
UserId field

Put the userid in the right place during a chat request so the gateway can track it.

mc-wright committed:
updated investigation prompt

mc-wright committed:
removed welcome message persistance for investigations

coreyogburn committed:
Split Models for Chat vs Session

Originally we were putting metadata for a session on the first message of that session. Now we have dedicated session objects in a new ES index.

A few frontend changes were made to adapt to these changes. Updated tests too.

Anybody running the assistant before this commit should remove the so-assistant index.

coreyogburn committed:
Fix Session Delete

mc-wright committed:
ai feature hiding logic implemented

mc-wright committed:
fixed partial json parsing, finished making onion ai a hidden feature, added option to restore last active chat upon navigation to chat page, fixed bug with brain icon not being grayed out

coreyogburn committed:
Remove Unused Fields and Defaults

coreyogburn committed:
Remove Binary, Update Gitignore

Never should have been included.

coreyogburn committed:
Replace Fetch with Axios

Found a way to make axios do streaming. Updated tests.

jertel commented on .gitignore:
```diff
@@ -19,3 +19,4 @@ __pycache__
 __debug_bin*
 .venv
 html/so-ca.crt
+securityonion-soc
```
What would cause the creation of a /securityonion-soc/securityonion-soc dir or file?

coreyogburn replied:
This was added after Josh B notified me I was about to merge in a 44MB binary with that name. It's the default name from a particular build command.

mc-wright committed:
most recent chats shown at the top instead of bottom

jertel commented on html/index.html:
```diff
@@ -225,6 +225,12 @@
               </template>
               <v-list-item-title v-text="i18n.home"></v-list-item-title>
             </v-list-item>
+            <v-list-item v-if="isLicensed('oai')" data-aid="nav_assistant" @click="" to="/assistant">
```
Doesn't this also need to check if it's enabled? We discussed this last week and I recall we don't want this showing up unless the admin enables it. The way it's currently implemented looks like it will always show up for any Pro license (so all licenses will have OAI available.

mc-wright replied:
fixed

jertel commented on html/js/app.js:
```diff
@@ -8,6 +8,9 @@ const routes = [];
 const components = [];
 const templatePromises = [];
 
+const FEAT_TTR = 'ttr';
```
Are these consts necessary? These are already defined below in the data map: 
<img width="370" height="159" alt="image" src="https://github.com/user-attachments/assets/132483ef-2be0-46fa-9220-3babee343c69" />

mc-wright replied:
fixed

jertel commented on html/pages/assistant.html:
```diff
@@ -0,0 +1,245 @@
+		display: flex;
+		min-height: 50vh;
+		justify-content: center;
+	">This feature is unavailable.</div>
```
Needs to be localized.

mc-wright replied:
fixed

jertel commented on html/pages/assistant.html:
```diff
@@ -0,0 +1,245 @@
+						</v-list-item>
+					</v-list>
+					<div v-else class="text-center text-medium-emphasis py-8">
+						<v-icon size="48" class="mb-4">fa-history</v-icon>
```
did none of the built-in sizes work? Is that why you had to hardcode a size here? In general I don't want UI sizes (fonts, icons, etc) hardcoded, since it goes against theme compatibility standards.

mc-wright replied:
I remember the built-in sizes not working for me when I tried to align the divider lines between the chat window and message history.

jertel commented on html/pages/assistant.html:
```diff
@@ -0,0 +1,245 @@
+												
+												<!-- Execution Time -->
+												<div v-if="toolUse.completedAt" class="text-caption text-medium-emphasis">
+													Completed: {{ formatTimestamp(toolUse.completedAt) }}
```
Needs localized

mc-wright replied:
fixed

jertel commented on html/pages/assistant.html:
```diff
@@ -0,0 +1,245 @@
+													<v-expansion-panels variant="accordion" class="tool-result-expansion">
+														<v-expansion-panel>
+															<v-expansion-panel-title class="text-caption" expand-icon="fa-angle-down" collapse-icon="fa-angle-up">
+																Raw Tool Result
```
Needs localized

mc-wright replied:
fixed

jertel commented on html/pages/assistant.html:
```diff
@@ -0,0 +1,245 @@
+															data-aid="reject_tool"
+														>
+															<v-icon size="small" class="mr-1">fa-times</v-icon>
+															Reject
```
Needs localized

mc-wright replied:
fixed

jertel commented on html/pages/assistant.html:
```diff
@@ -0,0 +1,245 @@
+															data-aid="approve_tool"
+														>
+															<v-icon size="small" class="mr-1">fa-check</v-icon>
+															Approve
```
Needs localized

mc-wright replied:
fixed

jertel commented on html/pages/assistant.html:
```diff
@@ -0,0 +1,245 @@
+															size="small"
+															variant="flat"
+															@click="approveTool(toolUse)"
+															data-aid="approve_tool"
```
data aid format: uicomponents_uniquename_type (type is optional and you can see when I've used it and when I haven't by looking at existing patterns). All data-aids need to be reviewed and made to follow the existing pattern.

mc-wright replied:
I made some changes to adhere to this formatting. Let me know if it looks better.

jertel commented on html/pages/assistant.html:
```diff
@@ -0,0 +1,245 @@
+												
+												<!-- Tool Approval Buttons -->
+												<div v-if="toolUse.status === 'pending_approval'" class="mb-2">
+													<div class="text-caption text-medium-emphasis mb-1">This tool wants to execute. Do you approve?</div>
```
Needs localized

mc-wright replied:
fixed

jertel commented on html/pages/assistant.html:
```diff
@@ -0,0 +1,245 @@
+												
+												<!-- Tool Parameters -->
+												<div v-if="Object.keys(toolUse.input).length > 0" class="mb-2">
+													<div class="text-caption text-medium-emphasis mb-1">Parameters:</div>
```
Needs localized

mc-wright replied:
fixed

jertel commented on html/pages/assistant.html:
```diff
@@ -0,0 +1,245 @@
+									<div class="d-flex align-center mb-1">
+										<span class="font-weight-medium text-body-2 mr-2">{{ message.role === 'user' ? 'You' : i18n.chat }}</span>
+										<span class="text-caption text-medium-emphasis mr-2">{{ formatTimestamp(message.timestamp) }}</span>
+										<span class="text-caption text-primary" v-if="message.usage && message.usage.credits">{{ message.usage.credits }} credits</span>
```
Needs localized

mc-wright replied:
fixed

jertel commented on html/pages/assistant.html:
```diff
@@ -0,0 +1,245 @@
+								</v-avatar>
+								<div class="flex-grow-1">
+									<div class="d-flex align-center mb-1">
+										<span class="font-weight-medium text-body-2 mr-2">{{ message.role === 'user' ? 'You' : i18n.chat }}</span>
```
Needs localized

mc-wright replied:
fixed

jertel commented on html/pages/assistant.html:
```diff
@@ -0,0 +1,245 @@
+		</v-col>
+		<v-spacer></v-spacer>
+		<v-col cols="12" lg="2" align="end">
+			<h4 :data-aid="'credits_remaining_' + 'chat'">{{ i18n.creditsRemaining }}: <span :class="creditsRemaining <= 500000 ? 'text-error' : 'text-primary'">{{ formatCount(creditsRemaining) }}</span></h4>
```
Should not be using hardcoded thresholds, these are supposed to be coming from configuration data.

mc-wright replied:
fixed

jertel commented on html/pages/assistant.html:
```diff
@@ -0,0 +1,245 @@
+								density="compact"
+								hide-details
+								class="my-0"
+								:label="i18n.restoreLastActive || 'Automatically restore last active chat session'"
```
We do not use fallback translations, please remove all of these.

mc-wright replied:
fixed

mc-wright replied:
fixed

jertel commented on html/pages/assistant.html:
```diff
@@ -0,0 +1,245 @@
+								<div class="flex-grow-1">
+									<div class="d-flex align-center mb-1">
+										<span class="font-weight-medium text-body-2 mr-2">AI Assistant</span>
+										<span class="text-caption text-medium-emphasis">typing...</span>
```
Needs localized

mc-wright replied:
fixed

jertel commented on html/pages/assistant.html:
```diff
@@ -0,0 +1,245 @@
+								</v-avatar>
+								<div class="flex-grow-1">
+									<div class="d-flex align-center mb-1">
+										<span class="font-weight-medium text-body-2 mr-2">AI Assistant</span>
```
Needs localized

mc-wright replied:
fixed

jertel commented on html/pages/assistant.html:
```diff
@@ -0,0 +1,245 @@
+									</div>
+									<div class="text-body-2">
+										<v-progress-linear indeterminate color="secondary" height="2"></v-progress-linear>
+										<span class="text-caption text-medium-emphasis">Thinking...</span>
```
Needs localized

mc-wright replied:
fixed

jertel reviewed (changes requested):
Review is incomplete.

coreyogburn committed:
Fixing Typing Issue in Unit Test

map[string]any was resulting in comparing ints to floats. Adding a decimal forces them to be float64s which is the expected type.

coreyogburn committed:
Fixed Typing Issue Properly

We're pulling a count from JS. All numbers are float64 in JS BUT it's a count, it'll always be a whole number. Convert it rather than compare floats that are really ints.

mc-wright committed:
added IDs to html

mc-wright committed:
fixed hidden feature enabledInSoc check in nav screen

mc-wright committed:
localization

mc-wright committed:
fixed data aids to adhere to formatting

mc-wright committed:
more data aid changes, cleaned up unused code

coreyogburn committed:
Data Streams

Use Data Streams instead of normal indexes for assistant storage. There's some ripples from this: sessions have a new field `sessionId` to represent the session instead of the default `id` because data streams won't let you specify doc Ids. This rippled all the way to the UI.

mc-wright committed:
cleaned up ai investigation code

mc-wright committed:
changed 'id' to 'sessionId' to account for Corey's changes

jertel committed:
Update .gitignore

jertel committed:
Update license_manager.go

jertel commented on model/assistant-chat.go:
```diff
@@ -0,0 +1,230 @@
+)
+
+type ChatRequest struct {
+	Messages      []*Message      `json:"messages"`
```
Are all of these structs exposed via the API? If so each field needs a comment and an example. See other structs to ensure consistency.

jertel commented on model/assistant-session.go:
```diff
@@ -0,0 +1,11 @@
+
+import "time"
+
+type AssistantSession struct {
```
Same comment about the API desc/example.

mc-wright committed:
threshold configs added for assistant context

jertel commented on model/clientparameters.go:
```diff
@@ -171,6 +173,11 @@ type AlertingParameters struct {
 }
 
+type AssistantParameters struct {
+	EnabledInSoc        bool   `json:"enabledInSoc"`
```
I'm curious why this is called InSoc. Why not just "Enabled"? Are you trying to differentiate UI vs server? Since this is already a clientParameter it's kind of implied.

jertel commented on model/tool.go:
```diff
@@ -0,0 +1,72 @@
+	"time"
+)
+
+type ToolRequest struct {
```
Same comment about the API desc/example. (If any of these are only used between the server and the cloud AI gateway, then ignore this)

jertel commented on server/modules/assistant/query_events.go:
```diff
@@ -0,0 +1,367 @@
+}
+
+func (t *QueryEventsTool) GetDescription() string {
+	return "- Execute OQL queries to retrieve Security Onion events\n" +
```
This would be easier to maintain as a backticked literal block

coreyogburn replied:
There are backticks throughout this string that make using backticked literal blocks even uglier than this approach.

jertel commented on server/modules/assistant/get_playbook_questions.go:
```diff
@@ -0,0 +1,205 @@
+
+	detId, ok := event.Payload["rule.uuid"].(string)
+	if !ok {
+		logger.WithField("event", event).Error("event does not have a rule.uuid field")
```
Make the event field more unique to avoid log storage collisions in ES.

coreyogburn committed:
Enforce Permissions in Store

Removed one unused function.

In addition to checking perms in the handlers, now checking assistant permissions in the assistant store.

jertel commented on html/js/routes/assistant.js:
```diff
@@ -0,0 +1,1322 @@
+  methods: {
+
+    initAssistant(params) {
+      console.log("assistant", params);
```
These console loggers need to be removed, or at most only use the `this.$root.log()` logger to log anomalies (but should be rarely needed since we want to show errors to the user, and not hide them in the console). When users open up dev tools I don't want anything showing up in the console output under a normal operational flow. If you look at all the other javascript route files you'll find no logging.

jertel commented on html/js/routes/assistant.js:
```diff
@@ -0,0 +1,1322 @@
+        this.messages = [
+          {
+            role: 'assistant',
+            content: 'Hello! I\'m your AI Assistant for Security Onion. How can I help you today?',
```
localize since this is not generated by the model

jertel commented on html/js/routes/assistant.js:
```diff
@@ -0,0 +1,1322 @@
+      if (!this.newMessage.trim()) return;
+      
+      if (!this.assistantEnabled) {
+        this.$root.showError('Onion AI is not enabled or is otherwise unavailable.');
```
localize

jertel commented on html/js/routes/assistant.js:
```diff
@@ -0,0 +1,1322 @@
+        // Show user-friendly error message
+        const errorMessage = {
+          role: 'assistant',
+          content: 'I apologize, but I\'m having trouble connecting to the AI service right now. Please try again in a moment.',
```
localize

jertel commented on html/js/routes/assistant.js:
```diff
@@ -0,0 +1,1322 @@
+// Copyright 2019 Jason Ertel (github.com/jertel).
```
In general, this file has too large of functions. They need to be broken apart to make the code easier to read and maintain. This can happen post .180.

jertel replied:
Added as .190 task

jertel commented on html/js/routes/hunt.js:
```diff
@@ -2338,6 +2391,81 @@ const huntComponent = {
+    async loadInvestigationSessions() {
+      // Load investigation sessions from backend
+      try {
+        const response = await this.$root.papi.get('/assistant/sessions');
```
Typically when we send an API request to the server we show a loading spinner to let the user know things are loading and about to change. Is there a reason this isn't done consistently in this PR, such as this location?

mc-wright replied:
This function is called in initHunt(), which is processed when the page is mounted, which all within the scope of a loading spinner.

jertel replied:
When the user clicks to load a different session does the spinner appear?

mc-wright replied:
All the investigation sessions are loaded when the user navigates to the alerts page (there is a spinner for this). If a user clicks on a brain, they are immediately navigated to the Onion AI page. It's unclear to me where we where a spinner would be needed, as the user is simply directed to a link (there is nothing to wait on).

mc-wright replied:
I added a loading spinner to various areas of the assistant page.

jertel commented on html/js/routes/hunt.js:
```diff
@@ -3176,20 +3304,139 @@ const huntComponent = {
+      if (existingInvestigation && existingInvestigation.chatSessionId) {
+        // Navigate to existing chat session
+        const chatUrl = `${window.location.origin}/#/assistant/${existingInvestigation.chatSessionId}`;
+        window.open(chatUrl, '_blank');
```
Is this intended to always open a new window/tab? If there's consistency with existing behavior on the Hunt screen that's fine but I know a lot of links don't assume the user wants a new tab, and instead the user can middle mouse click to open a new tab.

jertel replied:
Also, why window.open() vs $router.push/replace?

mc-wright replied:
The intention is to always open a new window, which intuitively made sense to me. I can change it to not always open a new window if you want.

jertel replied:
Yes, for consistency with the rest of the app, please have it open in the same tab.

mc-wright replied:
Okay, I made it so that it can be left clicked or middle clicked. I had to use a combination of $router.push and window.open to accomplish this.

jertel commented on html/js/routes/hunt.js:
```diff
@@ -3176,20 +3304,139 @@ const huntComponent = {
+        alertRule: item['rule.rule']
+      };
+
+      const investigationMsg = this.investigationMsg.replaceAll("{socid}", alertData.socId || 'Unknown');
```
Do we not have a method that already replaces vars? I thought we've done this in other parts of the app so if possible I don't want to have multiple code locations that are doing this.

mc-wright replied:
I wasn't aware of any methods like this, but I can take a look.

jertel replied:
Check with Corey - he was probably involved

mc-wright replied:
fixed

jertel reviewed (changes requested):
Still incomplete. New server go code is not yet reviewed.

mc-wright committed:
localization and console log cleanup

jertel commented on server/modules/assistant/helpers.go:
```diff
@@ -0,0 +1,64 @@
+)
+
+// parseTimeString parses various time formats used by Security Onion
+func parseTimeString(timeStr string) string {
```
It's surprising that we need more time formatters/parsers. I would think we have enough already in the codebase. Might be worthwhile to take an inventory of all Go time formatters/parsers because even if this is still needed perhaps it could be moved to a util pkg and rereplace other formatter/parser statements.

coreyogburn replied:
This parses relative time strings into a singular date. The only other place I found in Go parsing relative time strings similar to this is how in playbooks we parse out the dateRange field but it generates a range ("[date] - [date]") instead of a singular date. I'll move this and it's tests to the util package.

jertel commented on server/modules/assistant/helpers.go:
```diff
@@ -0,0 +1,64 @@
+			var duration time.Duration
+			switch unit {
+			case "m":
+				duration = time.Duration(amount) * time.Minute
```
What makes this routine only need day/hour/minute units but not seconds, weeks, months, or years?

coreyogburn replied:
The function was previously only used by a tool when the AI hunts for events. Seconds, weeks, and years can be easily added. Should months be an uppercase M to avoid conflicting with minutes?

jertel replied:
For now just omit Months since it will be confusing. I see there's a buildQuestionRange() function which is doing something similar to ParseRelativeTimeStr(). I didn't study it close enough to see if there's any reusability there, but perhaps that suffix detection logic could be shared.

coreyogburn committed:
Cleanup before Merge

enabledInSoc => enabled

Examples added to models' fields.

More specific log field names.

mc-wright committed:
more localization

jertel commented on server/modules/elastic/elasticassistantstore.go:
```diff
@@ -0,0 +1,533 @@
+
+	logger.WithFields(log.Fields{
+		"documentIndex": index,
+		"document":      store.truncate(document),
```
This log could be a problem since it appears to be logging the chat message. I might be misunderstanding what the document var represents though.

jertel replied:
I know we have some other debug statements that also log a truncated elastic response as well so perhaps we need to hide these behind a dev mode toggle.

jertel commented on server/modules/elastic/elasticassistantstore.go:
```diff
@@ -0,0 +1,533 @@
+	json, err := readJsonFromResponse(res)
+
+	logger.WithFields(log.Fields{
+		"response":  store.truncate(json),
```
Same issue here.

jertel commented on server/modules/elastic/elasticassistantstore.go:
```diff
@@ -0,0 +1,533 @@
+	}
+
+	logger.WithFields(log.Fields{
+		"response":  store.truncate(responseJSON),
```
Similar log sensitivity issue

jertel commented on server/modules/elastic/elasticassistantstore.go:
```diff
@@ -0,0 +1,533 @@
+	}
+
+	logger.WithFields(log.Fields{
+		"query":     store.truncate(string(queryJSON)),
```
Would be safe to rename this to "assistantEsQuery" to avoid collisions.

jertel commented on server/modules/elastic/elasticassistantstore.go:
```diff
@@ -0,0 +1,533 @@
+	}
+
+	logger.WithFields(log.Fields{
+		"response":  store.truncate(responseJSON),
```
Similar potential log sensitivity issue (and response key could be made more specific)

jertel replied:
This might be the last one. The response field should be elasticResponse if you decide it's necessary to keep it. Otherwise potentially change this to the length of the response.

jertel commented on server/modules/assistant/assistant.go:
```diff
@@ -0,0 +1,418 @@
+	}
+
+	logger.WithFields(log.Fields{
+		"result": result,
```
Rename result to something more specific. toolResult perhaps.

jertel reviewed (changes requested):
I think I've gone through everything at least once now.

coreyogburn committed:
Model Examples

Refactored TempBody into IncomingMessage.

Added example tags to several of our structs for documentation.

coreyogburn committed:
Field Descriptions

mc-wright committed:
router.push use for investigations, opens in same tab

mc-wright committed:
investigation button middle click functionality

mc-wright committed:
use replaceActionVar method

mc-wright committed:
fix

mc-wright committed:
loading spinner added to various operations in assistant page

mc-wright committed:
update assistant tests

mc-wright committed:
removed unused code for investigation feature

mc-wright committed:
fix

mc-wright committed:
removed unnecessary function call

coreyogburn committed:
Query Events now has a default of 10 if the limit isn't provided.

Cleaned up import order.

Added some more fields to the exclusion list of getExpandedData.

mc-wright committed:
fixed issue where the sessions api was being called twice when switching chats

mc-wright committed:
fix tests to account for last commit

coreyogburn committed:
Logs, Docs, and Util

Removed chat messages from logs in the store.

Added missing doc comments for ToolRequest. Improved log field names.

Refactored date parsing code into the util folder.

jertel commented on html/pages/assistant.html:
```diff
@@ -0,0 +1,249 @@
+				<v-card-title id="chat-header" class="py-3 d-flex align-center">
+					<v-icon class="mr-2">fa-comments</v-icon>
+					<span class="flex-grow-1">{{ i18n.chatTitle }}</span>
+					<div class="text-wrap" style="font-size: 14px;">Context Length: <span :class="getContextColor(contextLength)">{{ formatCount(contextLength) }}</span></div>
```
Needs localized

mc-wright replied:
fixed

jertel commented on html/js/routes/assistant.js:
```diff
@@ -0,0 +1,1255 @@
+      toolUse.error = this.i18n.assistantToolUseReject;
+      
+      // Add a message indicating the tool was rejected
+      const rejectionMessage = `Tool execution for "${toolUse.name}" was rejected by the user.`;
```
Localize?

mc-wright replied:
fixed

jertel commented on server/modules/assistant/get_playbook_questions.go:
```diff
@@ -0,0 +1,205 @@
+	return result, nil
+}
+
+type SimplePlaybook struct {
```
Are these structs only used between the SOC server and the AI model? If they're exposed in the API they will need descriptions and examples.

coreyogburn replied:
Those structs are only used in the conversation with the AI model.

mc-wright committed:
localization updates

coreyogburn committed:
Keeping Chats out of Logs

Fixes another log that should have response length instead of the raw response.

coreyogburn committed:
Consolidated Duration Parsing
